### PR TITLE
fix: [CDS-85942]: improving logs when manifest identifier is missing  in service identifiers

### PR DESCRIPTION
**Pull request by TarunUba1**
## Describe your changes
Reason: There was a case where customer was not sending identifier with manifest in serviceInputs which was throwing a bad exception, instead we will throw a general exception to prevent this scenario.
## Test

Tested for branch and others, for them it does not fail, hence doing it for only paths
After:
Screenshot 2023-12-13 at 4 43 57 PM

Before customer case:
Screenshot 2023-12-12 at 8 48 34 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56589)

### CODE REVIEW

**Comment by TarunUba1** [comment]
trigger uts

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java [hunk 1]
```diff
@@ -301,8 +301,11 @@ private void handleManifests(
         });
       }
     });
-
-    saveManifestsOutcome(ambiance, manifests, taskIdMapping);
+    try {
+      saveManifestsOutcome(ambiance, manifests, taskIdMapping);
+    } catch (Exception ex) {
+      throw new InvalidRequestException("File/Folder path must be present");
```
Comment (abhinavsingh1196) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:307]: could this be misleading during other exceptions?

Comment (TarunUba1) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:307]: saveManifestOutcome has one functionality which is to consume manifests into sweepingOutput, the other option is to add it in executionsweepingoutput was is but used everywhere,
The actual error is coming when we try to get identifier and map it butt its not there
RecastOrchestrationUtils.toJson(value) throws this error actually

Comment (abhinavsingh1196) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:307]: Also, we should probably throw hint/explanation exception here. I feel that this message by itself is not enough

**Comment by abhinavsingh1196** [comment]
Small comment above, please check.

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java (hunk 1 above)
Comment (abosii-harness) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:307]: This doesn't work

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java (hunk 1 above)
Comment (TarunUba1) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java]: yeah i cross checked the YAML of customer then realized it was for service input

**Comment by TarunUba1** [comment]
trigger uts

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/mappers/ManifestOutcomeValidator.java [hunk 2]
```diff
@@ -283,11 +284,15 @@ private boolean hasValue(ParameterField<String> parameterField, boolean allowExp
...
-      List<String> stringList = getParameterFieldValue(parameterField);
-      return stringList.stream().noneMatch(StringUtils::isBlank);
+    try {
+      if (!ParameterField.isNull(parameterField) && isNotEmpty(getParameterFieldValue(parameterField))) {
+        List<String> stringList = getParameterFieldValue(parameterField);
+        return stringList.stream().noneMatch(StringUtils::isBlank);
+      }
+      return false;
+    } catch (Exception ex) {
+      throw new InvalidRequestException("File/Folder path must be present");
```
Comment (abosii-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/mappers/ManifestOutcomeValidator.java:294]: Can we make this error a bit more generic, we could actually do the next thing:
- Wrap hasValue with another method that will have as input parameterField, fieldName and manifestId
- Once the exception is caught, we can provide the next error message:
Incorrect value {parameterField.getValue()} for field {fieldName} in manifest {manifestId}. 

With some hints like:
- Check provided value for {fieldName} in manifest {manifestId}
-  If using runtime input make sure that valid input is provided for manifest with {manifestId}. Input yaml should contain identifier {manifestId}

Comment (TarunUba1) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/mappers/ManifestOutcomeValidator.java]: i have added  a hint, couldn't introduce NestedException cause it was not working correctly.

**Comment by TarunUba1** [comment]
trigger uts

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/mappers/ManifestOutcomeValidator.java [hunk 3]
```diff
@@ -49,6 +50,9 @@
 @OwnedBy(CDP)
 @UtilityClass
 public class ManifestOutcomeValidator {
+  private final String SERVICE_INPUTS_MANIFEST_FAILURE = "Incorrect value [%s] for field [%s] in manifest [%s] \n"
+      + "Hint: Please check provided field [%s] for manifest id [%s]."
+      + "Valid input with an identifier is required in the YAML file for proper manifest processing during runtime.";
```
Comment (abosii-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/mappers/ManifestOutcomeValidator.java:55]: please add a space either after . or before Valid
Also this message may be confusing, let's make sure that hint is applicable in context of runtime input

**Review by GoelPratyush99 — APPROVED** [review]

**Comment by TarunUba1** [comment]
trigger uts

**Comment by TarunUba1** [comment]
trigger ut6, ut7, ut8

**Comment by TarunUba1** [comment]
trigger uts

**Review by abosii-harness — APPROVED** [review]

**Comment by TarunUba1** [comment]
trigger review, messagecheck

**Pull request merged**